lib/Sema/DebuggerTestingTransform.cpp

```diff
@@ -0,0 +1,245 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file This transform inserts instrumentation which the debugger can use to
+/// test its expression evaluation facilities.
+///
+//===----------------------------------------------------------------------===//
+
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTNode.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/DeclContext.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/Stmt.h"
+#include "swift/Subsystems.h"
+
+#include "TypeChecker.h"
+
+using namespace swift;
+
+namespace {
+
+/// Find available closure discriminators.
+///
+/// The parser typically takes care of assigning unique discriminators to
+/// closures, but the parser is unavailable to this transform.
+class DiscriminatorFinder : public ASTWalker {
+  unsigned NextDiscriminator = 0;
+
+public:
+  Expr *walkToExprPost(Expr *E) override {
+    auto *ACE = dyn_cast<AbstractClosureExpr>(E);
+    if (!ACE)
+      return E;
+
+    unsigned Discriminator = ACE->getDiscriminator();
+    assert(Discriminator != AbstractClosureExpr::InvalidDiscriminator &&
+           "Existing closures should have valid discriminators");
+    if (Discriminator >= NextDiscriminator)
+      NextDiscriminator = Discriminator + 1;
+    return E;
+  }
+
+  // Get the next available closure discriminator.
+  unsigned getNextDiscriminator() {
+    if (NextDiscriminator == AbstractClosureExpr::InvalidDiscriminator)
+      llvm::report_fatal_error("Out of valid closure discriminators");
+    return NextDiscriminator++;
+  }
+};
+
+/// Instrument decls with sanity-checks which the debugger can evaluate.
+class DebuggerTestingTransform : public ASTWalker {
+  ASTContext &Ctx;
+  DiscriminatorFinder &DF;
+  std::vector<DeclContext *> LocalDeclContextStack;
+
+public:
+  DebuggerTestingTransform(ASTContext &Ctx, DiscriminatorFinder &DF)
+      : Ctx(Ctx), DF(DF) {}
+
+  bool walkToDeclPre(Decl *D) override {
+    pushLocalDeclContext(D);
+
+    // Whitelist the kinds of decls to transform.
+    // TODO: Expand the set of decls visited here.
+    if (auto *FD = dyn_cast<AbstractFunctionDecl>(D))
+      return !FD->isImplicit() && FD->getBody();
+    else if (auto *TLCD = dyn_cast<TopLevelCodeDecl>(D))
+      return !TLCD->isImplicit() && TLCD->getBody();
+    return false;
+  }
+
+  bool walkToDeclPost(Decl *D) override {
+    popLocalDeclContext(D);
+    return true;
+  }
+
+  std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+    pushLocalDeclContext(E);
+
+    // Whitelist the kinds of exprs to transform.
+    // TODO: Expand the set of exprs visited here.
+    if (auto *AE = dyn_cast<AssignExpr>(E))
```

Comment:
Eventually we might want to move this to an helper, i.e. `shouldTransform()`, but I think this is premature for now.